Move some utility functions out of `util` and into their own packages

kminehart opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with master
-->

**What this PR does**:

This PR reduces the number of imports in the `pkg/chunk/cache` for use outside of cortex.

When importing `github.com/cortexproject/cortex/pkg/chunk/cache` I was getting a lot of extra dependencies because `cache` imports `util` which imports packages like `grpc-go`.

This involved moving functions out of `util` and into their own packages.

To maintain compatibility and reduce the number of changes, I've left the functions in `util`, and had them call their equivalent function in their new package. If we don't want the cruft of having these defined twice, I'd be happy to comb through and refactor it all.


**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

jtlisi reviewed (commented):
> To maintain compatibility and reduce the number of changes, I've left the functions in util, and had them call their equivalent function in their new package. If we don't want the cruft of having these defined twice, I'd be happy to comb through and refactor it all.

I was about to comment about this, but then I re-read you description. I do think it would make sense to remove the original files and to update everywhere that calls these functions. It would also make the diff a bit cleaner since a lot of the files should be a 1-1 copy over besides the package name references.

pstibrany commented on pkg/util/ctxutil/ctx.go:
```diff
@@ -0,0 +1 @@
+package ctxutil
```
Why do we need this empty package?

kminehart replied:
I put `WithContext` in here, but then I realized it needed to be grouped with logging functions. I must have forgotten to remove it.I'll remove this empty package.

pstibrany commented on pkg/util/logutil/warn.go:
```diff
@@ -0,0 +1,21 @@
+package logutil
```
More like `experimental.go`

pstibrany commented on pkg/util/logutil/log.go:
```diff
@@ -0,0 +1,7 @@
+package logutil
```
I'd suggest not repeating "util" suffix in these package names, and only use non-conflicting names on imports.

kminehart replied:
What name would you suggest in this case?

pstibrany replied:
`pkg/util/log`, and import using `import util_log "github.com/cortexproject/cortex/pkg/util/log"`.

pstibrany commented on pkg/util/experimental.go:
```diff
@@ -1,21 +1,8 @@
+import "github.com/cortexproject/cortex/pkg/util/logutil"
 
 // WarnExperimentalUse logs a warning and increments the experimental features metric.
 func WarnExperimentalUse(feature string) {
```
Alternative:

```
var WarnExperimentalUse = logutil.WarnExperimentalUse
```

I don't think we even need this. Let's just update the references.

kminehart replied:
hahahaha good point. I  completely forgot about this.

pstibrany commented on pkg/util/log.go:
```diff
@@ -9,16 +9,15 @@ import (
 	// Logger is a shared go-kit logger.
 	// TODO: Change all components to take a non-global logger via their constructors.
-	Logger = log.NewNopLogger()
+	Logger = logutil.Logger
```
If we want to keep this var to avoid lot of code changes, let's at least mark it as deprecated.

pstibrany commented on pkg/util/math.go:
```diff
@@ -1,33 +1,25 @@
 package util
 
+import (
```
There are only handful references for each of these functions, I'd suggest replacing them with new functions and deleting this file.

pstibrany reviewed (commented):
> I was about to comment about this, but then I re-read you description. I do think it would make sense to remove the original files and to update everywhere that calls these functions. It would also make the diff a bit cleaner since a lot of the files should be a 1-1 copy over besides the package name references.

I agree with Jacob that we should just remove old functions and variables. Note however that we should then also update  projects that use this (eg. Loki, Thanos), to avoid the pain for them.

kminehart commented:
* Renamed logutil/mathutil to log/math
* Removed old definitions in `util` and replaced references to these functions throughout the codebase
* When a conflict arose (pretty common with these 2 package names), I went with `util_log` or `util_math` to show that they were subpackages of the `util` package.

pracucci commented on pkg/alertmanager/api.go:
```diff
@@ -9,7 +9,7 @@ import (
 	"github.com/cortexproject/cortex/pkg/alertmanager/alerts"
 	"github.com/cortexproject/cortex/pkg/tenant"
-	"github.com/cortexproject/cortex/pkg/util"
+	util_log "github.com/cortexproject/cortex/pkg/util/log"
```
I think the `pkg/util/log` should be renamed to `pkg/util/logutil` to avoid misunderstandings with `github.com/go-kit/kit/log`.

kminehart replied:
that's what they were originally; @pstibrany requested that I rename them :see_no_evil:

pstibrany replied:
> I think the `pkg/util/log` should be renamed to `pkg/util/logutil` to avoid misunderstandings with `github.com/go-kit/kit/log`.

:) In my previous comment I actually suggested we do the opposite, to avoid repeating “util” in the package name. I don’t mind the conflicting names, and actually prefer “util_log” and “util_math” in imports (I wouldn’t mind to use that everywhere for consistency, but there is no need to).

pracucci replied:
Didn't know you already discussed it. I missed that, sorry. Let's discuss it offline, but definitely not a good reason to block this PR.

pracucci commented on pkg/ingester/ingester_v2_test.go:
```diff
@@ -41,6 +41,7 @@ import (
 	"github.com/cortexproject/cortex/pkg/ring"
 	cortex_tsdb "github.com/cortexproject/cortex/pkg/storage/tsdb"
 	"github.com/cortexproject/cortex/pkg/util"
+	util_math "github.com/cortexproject/cortex/pkg/util/math"
```
`pkg/util/math` should be renamed to `pkg/util/mathutil` for the same reason.

pracucci commented on pkg/util/log/log.go:
```diff
@@ -0,0 +1,27 @@
+)
+
+var (
+	Logger = kitlog.NewNopLogger()
```
This should be marked deprecated too.

pracucci reviewed (commented):
Thanks for working on this! I agree with the changes, but let's rename the util packages for clarity (and to avoid pkg naming clashes):

- `pkg/util/log` > `pkg/util/logutil`
- `pkg/util/math` > `pkg/util/mathutil`
